pep-0718.rst

```diff
@@ -0,0 +1,152 @@
+Topic: Typing
+Content-Type: text/x-rst
+Created: 23-Jun-2023
+Python-Version: 3.13
+Post-History: `24-Jun-2023 <https://discuss.python.org/t/28457/>`__
+
+Abstract
+--------
+
+This PEP proposes making ``FunctionType``\ , ``MethodType``\ , ``BuiltinFunctionType``
+and ``BuiltinMethodType`` instances subscriptable for typing purposes. Doing so allows
+for specialisation in complex, hard-to-infer situations for type checkers where
+bi-directional inference (which allows for the types of parameters of anonymous
+functions to be inferred) is insufficient.
+
+Motivation
+----------
+
+Currently, it is not possible to infer certain situations for generic functions where
+unspecialised literals are involved
+
+.. code-block:: python
+
+   def make_list[T](*args: T) -> list[T]: ...
+   reveal_type(make_list())  # type checker cannot infer T
+
+Making instances of ``FunctionType`` subscriptable would allow for this constructor to
+be typed
+
+.. code-block:: python
+
+   reveal_type(make_list[int]())  # type is list[int]
+
+Currently you have to use an assignment to allow for this to be type-able
+
+.. code-block:: python
+
+   x: list[int] = make_list()
+   reveal_type(x)  # type is list[int]
+
+but this code is unnecessarily verbose taking up multiple lines for a simple function
+call.
+
+.. code-block:: python
+
+   def factory[T](func: Callable[[T], Any]) -> Foo[T]: ...
+
+   reveal_type(factory(lambda x: "Hello World" * x))  # type checker cannot infer T (again)
+
+If functions were subscriptable, however, the following could be performed
+
+.. code-block:: python
+
+   reveal_type(factory[int](lambda x: "Hello World" * x))  # type is Foo[int]
+
+Currently, with unspecialised literals, it is not possible to determine a type for
+situations similar to:
+
+.. code-block:: python
+
+   def foo[T](x: list[T]) -> T: ...
+   reveal_type(foo([]))  # type checker cannot infer T (yet again)
+
+.. code-block:: python
+
+   reveal_type(foo[int]([]))  # type is int
+
+It is also useful to be able to specify in cases in which a certain type must be passed
+to a function beforehand
+
+.. code-block:: python
+
+   words = ["hello", "world"]
+   foo[int](words)  # Invalid: list[str] is incompatible with list[int]
+
+Allowing subscription makes functions and methods consistent with generic classes where
+they weren't already. Whilst all of the proposed changes can be implemented using
+callable generic classes, syntactic sugar would be highly welcome.
+
+Due to this, specialising the function and using it as a new factory is fine
+
+.. code-block:: python
+
+   make_int_list = make_list[int]
+   reveal_type(make_int_list())  # type is list[int]
+
+This proposal also opens the door to
+`monomorphisation <https://en.wikipedia.org/wiki/Monomorphization>`_ and
+`reified types <https://en.wikipedia.org/wiki/Reification_(computer_science)>`_
+
+Specification
+-------------
+
+``FunctionType``\ , ``MethodType``\ , ``BuiltinFunctionType`` and ``BuiltinMethodType``
```

Comment:
I don't think there's much point considering it's completely opt in but I can try and let the faster cpython team know

Reply:
The sponsor of this PEP is on the faster CPython team so it's safe to assume they know about this.

I have been thinking about this too. Subscripted calls are likely to be quite a bit slower than normal calls, because they would involve calling `FunctionType.__getitem__`, allocating and initializing a `GenericAlias` object, and calling its `tp_call` method, which then calls the actual function. That's going to be quite a bit slower than a straight Python-to-Python call.

Does that matter? Maybe enough to be noticeable if you frequently call explicitly specialized generic functions in a performance-critical part of your code. However, if this turns out to be a real-world problem, we can likely blunt most of the impact through PEP-659 specialization.